File: src/python/xerxes/runtime/bootstrap.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Bootstrap module for Xerxes.

Exports:
    - BootstrapStage
    - BootstrapResult
    - bootstrap"""

from __future__ import annotations

import platform
import subprocess
import sys
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any

from .execution_registry import ExecutionRegistry


@dataclass
class BootstrapStage:
    """Bootstrap stage.

    Attributes:
        name (str): name.
        status (str): status.
        detail (str): detail.
        duration_ms (float): duration ms."""

    name: str
    status: str = "ok"
    detail: str = ""
    duration_ms: float = 0.0


@dataclass
class BootstrapResult:
    """Bootstrap result.

    Attributes:
        stages (list[BootstrapStage]): stages.
        registry (ExecutionRegistry): registry.
        system_prompt (str): system prompt.
        context (dict[str, Any]): context."""

    stages: list[BootstrapStage] = field(default_factory=list)
    registry: ExecutionRegistry = field(default_factory=ExecutionRegistry)
    system_prompt: str = ""
    context: dict[str, Any] = field(default_factory=dict)

    @property
    def ok(self) -> bool:
        """Return Ok.

        Args:
            self: IN: The instance. OUT: Used for attribute access.
        Returns:
            bool: OUT: Result of the operation."""
        return all(s.status != "failed" for s in self.stages)

    def as_markdown(self) -> str:
        """As markdown.

        Args:
            self: IN: The instance. OUT: Used for attribute access.
        Returns:
            str: OUT: Result of the operation."""
        lines = [
            "# Bootstrap Report",
            "",
            f"Status: {'OK' if self.ok else 'FAILED'}",
            f"Stages: {len(self.stages)}",
            "",
        ]
        for stage in self.stages:
            icon = {"ok": "+", "skipped": "~", "failed": "!"}[stage.status]
            lines.append(f"- [{icon}] {stage.name}: {stage.detail} ({stage.duration_ms:.1f}ms)")
        return "\n".join(lines)


def bootstrap(
    model: str = "",
    cwd: str | Path | None = None,
    tools: list[Any] | None = None,
    commands: dict[str, Any] | None = None,
    include_git_info: bool = True,
    include_xerxes_md: bool = True,
    extra_context: str = "",
) -> BootstrapResult:
    """Bootstrap.

    Args:
        model (str, optional): IN: model. Defaults to ''. OUT: Consumed during execution.
        cwd (str | Path | None, optional): IN: cwd. Defaults to None. OUT: Consumed during execution.
        tools (list[Any] | None, optional): IN: tools. Defaults to None. OUT: Consumed during execution.
        commands (dict[str, Any] | None, optional): IN: commands. Defaults to None. OUT: Consumed during execution.
        include_git_info (bool, optional): IN: include git info. Defaults to True. OUT: Consumed during execution.
        include_xerxes_md (bool, optional): IN: include xerxes md. Defaults to True. OUT: Consumed during execution.
        extra_context (str, optional): IN: extra context. Defaults to ''. OUT: Consumed during execution.
    Returns:
        BootstrapResult: OUT: Result of the operation."""

    import time

    result = BootstrapResult()
    working_dir = Path(cwd) if cwd else Path.cwd()

    t0 = time.monotonic()
    result.context = {
        "cwd": str(working_dir),
        "python_version": f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
        "platform": platform.platform(),
        "model": model,
        "date": datetime.now().strftime("%Y-%m-%d %A"),
    }
    result.stages.append(
        BootstrapStage(
            name="environment",
            detail=f"Python {result.context['python_version']} on {platform.system()}",
            duration_ms=(time.monotonic() - t0) * 1000,
        )
    )

    t0 = time.monotonic()
    git_info = ""
    if include_git_info:
        git_info = _get_git_info(working_dir)
        result.context["git_info"] = git_info
    result.stages.append(
        BootstrapStage(
            name="git_info",
            status="ok" if git_info else "skipped",
            detail=git_info[:80] if git_info else "Not a git repository",
            duration_ms=(time.monotonic() - t0) * 1000,
        )
    )

    t0 = time.monotonic()
    xerxes_md = ""
    if include_xerxes_md:
        xerxes_md = _load_xerxes_md(working_dir)
        result.context["xerxes_md"] = xerxes_md
    result.stages.append(
        BootstrapStage(
            name="xerxes_md",
            status="ok" if xerxes_md else "skipped",
            detail=f"{len(xerxes_md)} chars" if xerxes_md else "No XERXES.md found",
            duration_ms=(time.monotonic() - t0) * 1000,
        )
    )

    t0 = time.monotonic()
    if commands:
        for name, handler in commands.items():
            result.registry.register_command(name, handler=handler)
    for cmd_name in [
        "help",
        "clear",
        "history",
        "save",
        "load",
        "model",
        "config",
        "cost",
        "context",
        "memory",
        "agents",
        "skills",
    ]:
        if not result.registry.get_command(cmd_name):
            result.registry.register_command(cmd_name, description=f"/{cmd_name} command")
    result.stages.append(
        BootstrapStage(
            name="commands",
            detail=f"{result.registry.command_count} commands registered",
            duration_ms=(time.monotonic() - t0) * 1000,
        )
    )

    t0 = time.monotonic()
    if tools:
        result.registry.register_from_agent_functions(tools)
    result.stages.append(
        BootstrapStage(
            name="tools",
            detail=f"{result.registry.tool_count} tools registered",
            duration_ms=(time.monotonic() - t0) * 1000,
        )
    )

    t0 = time.monotonic()
    result.system_prompt = _build_system_prompt(result.context, extra_context)
    result.stages.append(
        BootstrapStage(
            name="system_prompt",
            detail=f"{len(result.system_prompt)} chars",
            duration_ms=(time.monotonic() - t0) * 1000,
        )
    )

    return result


def _get_git_info(cwd: Path) -> str:
    """Internal helper to get git info.

    Args:
        cwd (Path): IN: cwd. OUT: Consumed during execution.
    Returns:
        str: OUT: Result of the operation."""

    try:
        branch = subprocess.check_output(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            cwd=str(cwd),
            stderr=subprocess.DEVNULL,
            text=True,
        ).strip()
        status = subprocess.check_output(
            ["git", "status", "--short"],
            cwd=str(cwd),
            stderr=subprocess.DEVNULL,
            text=True,
        ).strip()
        log = subprocess.check_output(
            ["git", "log", "--oneline", "-5"],
            cwd=str(cwd),
            stderr=subprocess.DEVNULL,
            text=True,
        ).strip()

        parts = [f"Branch: {branch}"]
        if status:
            parts.append(f"Status:\n{status}")
        if log:
            parts.append(f"Recent commits:\n{log}")
        return "\n".join(parts)
    except Exception:
        return ""


def _load_xerxes_md(cwd: Path) -> str:
    """Internal helper to load xerxes md.

    Args:
        cwd (Path): IN: cwd. OUT: Consumed during execution.
    Returns:
        str: OUT: Result of the operation."""

    from xerxes.core.paths import xerxes_subdir
    from xerxes.security.prompt_scanner import scan_context_content

    parts = []

    global_md = xerxes_subdir("XERXES.md")
    if global_md.exists():
        try:
            raw = global_md.read_text(encoding="utf-8")
            safe = scan_context_content(raw, filename="Global XERXES.md")
            parts.append(f"[Global XERXES.md]\n{safe}")
        except Exception:
            pass

    p = cwd
    for _ in range(10):
        candidate = p / "XERXES.md"
        if candidate.exists():
            try:
                raw = candidate.read_text(encoding="utf-8")
                safe = scan_context_content(raw, filename=f"Project XERXES.md: {candidate}")
                parts.append(f"[Project XERXES.md: {candidate}]\n{safe}")
            except Exception:
                pass
            break
        parent = p.parent
        if parent == p:
            break
        p = parent

    return "\n\n".join(parts)


def _build_system_prompt(context: dict[str, Any], extra: str = "") -> str:
    """Internal helper to build system prompt.

    Args:
        context (dict[str, Any]): IN: context. OUT: Consumed during execution.
        extra (str, optional): IN: extra. Defaults to ''. OUT: Consumed during execution.
    Returns:
        str: OUT: Result of the operation."""

    parts = [
        "You are Xerxes, an AI coding assistant with access to tools via function calling.",
        "",
        "# Tools",
        "ReadFile, WriteFile, FileEditTool, AppendFile, ListDir, GlobTool, GrepTool,",
        "ExecuteShell, ExecutePythonCode, GoogleSearch, DuckDuckGoSearch, WebScraper,",
        "AgentTool, SendMessageTool, TodoWriteTool, Calculator, JSONProcessor, CSVProcessor",
        "",
        "# How to decide",
        "1. Can you answer from knowledge alone? → Reply directly.",
        "2. Need to read a file? → ReadFile(file_path=...)",
        "3. Need to write a file? → WriteFile(file_path=..., content=...)",
        "4. Need to edit a file? → FileEditTool(file_path=..., old_string=..., new_string=...)",
        "5. Need to run a command? → ExecuteShell(command=...)",
        "6. Need to find files? → GlobTool(pattern=...)",
        "7. Need to search code? → GrepTool(pattern=...)",
        "8. Need web info? → GoogleSearch(query=...)  (preferred — uses Google CSE API when GOOGLE_API_KEY+GOOGLE_CSE_ID are set; otherwise scrapes google.com).",
        "   Or → DuckDuckGoSearch(query=...) when Google blocks the scrape and no API key is set.",
        "9. Need to list a directory? → ListDir(directory_path=...)",
        "",
        "# Web search via curl (when GoogleSearch returns 0 results)",
        "Google blocks naked scrapes from datacenter IPs. When GoogleSearch comes back empty,",
        "fall back to ExecuteShell with a real browser User-Agent.",
        "",
        "RECIPE 1 — Google raw HTML:",
        "  curl -sSL --compressed -A 'Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15' -H 'Accept-Language: en-US,en;q=0.9' 'https://www.google.com/search?q=YOUR+QUERY&num=10&hl=en'",
        "",
        "RECIPE 2 — Google + extract URLs in one pipe:",
        "  curl -sSL --compressed -A 'Mozilla/5.0' 'https://www.google.com/search?q=YOUR+QUERY&num=10' | grep -oE 'href=\"https?://[^\"]+' | grep -vE 'google\\.|youtube\\.|webcache' | sort -u | head -20",
        "",
        "RECIPE 3 — DuckDuckGo HTML (never blocks; use when Google returns the JS bot stub):",
        "  curl -sSL --compressed -A 'Mozilla/5.0' 'https://html.duckduckgo.com/html/?q=YOUR+QUERY'",
        "",
        "Tips:",
        "  - URL-encode spaces as '+' (or %20)",
        "  - Add '&num=N' to control result count (max 30)",
        "  - Add '&tbs=qdr:d|w|m|y' for day/week/month/year recency",
        "  - For site-specific: prepend 'site:example.com' to the query",
        "  - If Google returns ~90KB of JS with no <h3> tags, that's the bot stub — retry RECIPE 3",
        "",
        "# Multi-Agent Orchestration (CRITICAL)",
        "You have access to sub-agent spawning tools. Use them AGGRESSIVELY.",
        "- Default to PARALLEL execution. Sequential work is a last resort.",
        "- If a task has 2+ independent parts → SpawnAgents with wait=true.",
        "- If a task has 3+ parts → SpawnAgents is MANDATORY, not optional.",
        "- Research + implementation + review → 3 parallel agents, not 1 monolithic response.",
        "- Writing code in multiple files → SpawnAgents with coder subagents + worktree isolation.",
        "- Do NOT write all the code yourself. Delegate to coder sub-agents.",
        "- AgentTool: spawn one agent. SpawnAgents: spawn many in parallel.",
        "- TaskCreateTool: fire-and-forget background work.",
        "- Always name your agents descriptively so you can reference them later.",
        "",
        "# Critical",
        "- Be concise and direct.",
        "- Read files before editing them.",
        "- Use absolute paths for file operations.",
        "",
        "# Environment",
        f"- Date: {context.get('date', '')}",
        f"- CWD: {context.get('cwd', '')}",
        f"- Platform: {context.get('platform', '')}",
        f"- Model: {context.get('model', '')}",
    ]

    git_info = context.get("git_info", "")
    if git_info:
        parts.extend(["", "# Git", git_info])

    xerxes_md = context.get("xerxes_md", "")
    if xerxes_md:
        parts.extend(["", "# Project Context", xerxes_md])

    if extra:
        parts.extend(["", extra])

    return "\n".join(parts)


__all__ = [
    "BootstrapResult",
    "BootstrapStage",
    "bootstrap",
]

File: src/typescript/src/runtime/bootstrap.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { readFile } from 'node:fs/promises'
import { dirname, join, resolve } from 'node:path'

import { xerxesSubdirFor } from '../core/paths.js'
import { scanContextContent } from '../security/promptScanner.js'
import { ExecutionRegistry, type EntryHandler } from './executionRegistry.js'
import { loadProjectAgentWorkspace } from './projectWorkspace.js'

export const DEFAULT_BOOTSTRAP_COMMANDS = [
  'help',
  'clear',
  'history',
  'save',
  'load',
  'model',
  'config',
  'cost',
  'context',
  'memory',
  'agents',
  'skills'
] as const

/** Maximum UTF-8 payload reserved for the changed-file listing in bootstrap context. */
export const MAX_BOOTSTRAP_GIT_STATUS_BYTES = 8 * 1024

/** Injectable Git command boundary used by the system bootstrap and focused tests. */
export type GitCommandRunner = (arguments_: readonly string[], cwd: string) => Promise<string>

export type BootstrapStageStatus = 'failed' | 'ok' | 'skipped'

export interface BootstrapStage {
  readonly detail: string
  readonly durationMs: number
  readonly name: string
  readonly status: BootstrapStageStatus
}

export interface BootstrapHost {
  readonly cwd: () => string
  readonly date: () => Date
  readonly gitInfo: (cwd: string) => Promise<string>
  readonly monotonicNow: () => number
  readonly platform: () => string
  readonly projectWorkspace: (cwd: string) => Promise<string>
  readonly readText: (path: string) => Promise<string | undefined>
  readonly runtimeVersion: () => string
  readonly xerxesHomeFile: (name: string) => string
}

export interface BootstrapOptions {
  readonly commands?: Readonly<Record<string, EntryHandler | undefined>>
  readonly cwd?: string
  readonly extraContext?: string
  readonly host?: BootstrapHost
  readonly includeGitInfo?: boolean
  readonly includeXerxesMd?: boolean
  readonly model?: string
  readonly tools?: readonly unknown[]
}

export interface BootstrapResultOptions {
  readonly context: Readonly<Record<string, string>>
  readonly registry: ExecutionRegistry
  readonly stages: readonly BootstrapStage[]
  readonly systemPrompt: string
}

/** Aggregate output of the native session-bootstrap pipeline. */
export class BootstrapResult {
  readonly context: Readonly<Record<string, string>>
  readonly registry: ExecutionRegistry
  readonly stages: readonly BootstrapStage[]
  readonly systemPrompt: string

  constructor(options: BootstrapResultOptions) {
    this.context = Object.freeze({ ...options.context })
    this.registry = options.registry
    this.stages = Object.freeze([...options.stages])
    this.systemPrompt = options.systemPrompt
  }

  get ok(): boolean {
    return !this.stages.some(stage => stage.status === 'failed')
  }

  /** Render a small diagnostic report suitable for `/doctor` or a startup log. */
  asMarkdown(): string {
    const lines = [
      '# Bootstrap Report',
      '',
      'Status: ' + (this.ok ? 'OK' : 'FAILED'),
      'Stages: ' + this.stages.length,
      ''
    ]
    for (const stage of this.stages) {
      const icon = stage.status === 'ok' ? '+' : stage.status === 'skipped' ? '~' : '!'
      lines.push(`- [${icon}] ${stage.name}: ${stage.detail} (${stage.durationMs.toFixed(1)}ms)`)
    }
    return lines.join('\n')
  }
}

/**
 * Assemble startup context, the execution registry, and a safe default system prompt.
 *
 * The host owns all filesystem, git, clock, and project-workspace effects. Failed
 * optional context probes become explicit skipped stages rather than stopping a turn.
 */
export async function bootstrap(options: BootstrapOptions = {}): Promise<BootstrapResult> {
  const host = options.host ?? createSystemBootstrapHost()
  const registry = new ExecutionRegistry()
  const stages: BootstrapStage[] = []
  const environmentStarted = host.monotonicNow()
  const workingDirectory = resolve(options.cwd ?? host.cwd())
  const context: Record<string, string> = {
    cwd: workingDirectory,
    runtime_version: host.runtimeVersion(),
    platform: host.platform(),
    model: options.model ?? '',
    date: formatBootstrapDate(host.date())
  }

  stages.push(
    stageFromElapsed(host, 'environment', 'ok', `${context.runtime_version} on ${context.platform}`, environmentStarted)
  )

  const includeGitInfo = options.includeGitInfo ?? true
  const gitStarted = host.monotonicNow()
  const gitInfo = includeGitInfo ? await optionalContext(() => host.gitInfo(workingDirectory)) : ''
  if (includeGitInfo) context.git_info = gitInfo
  stages.push(
    stageFromElapsed(
      host,
      'git_info',
      gitInfo ? 'ok' : 'skipped',
      gitInfo ? gitInfo.slice(0, 80) : includeGitInfo ? 'Not a git repository' : 'Disabled',
      gitStarted
    )
  )

  const includeXerxesMd = options.includeXerxesMd ?? true
  const xerxesMdStarted = host.monotonicNow()
  const xerxesMd = includeXerxesMd ? await loadXerxesMd(workingDirectory, host) : ''
  if (includeXerxesMd) context.xerxes_md = xerxesMd
  stages.push(
    stageFromElapsed(
      host,
      'xerxes_md',
      xerxesMd ? 'ok' : 'skipped',
      xerxesMd ? xerxesMd.length + ' chars' : includeXerxesMd ? 'No XERXES.md found' : 'Disabled',
      xerxesMdStarted
    )
  )

  const projectWorkspaceStarted = host.monotonicNow()
  const projectWorkspace = await optionalContext(() => host.projectWorkspace(workingDirectory))
  context.project_agent_workspace = projectWorkspace
  stages.push(
    stageFromElapsed(
      host,
      'project_agent_workspace',
      projectWorkspace ? 'ok' : 'skipped',
      projectWorkspace ? projectWorkspace.length + ' chars' : 'No .agents workspace found',
      projectWorkspaceStarted
    )
  )

  const commandsStarted = host.monotonicNow()
  for (const [name, handler] of Object.entries(options.commands ?? {})) {
    registry.registerCommand(name, handler)
  }
  for (const name of DEFAULT_BOOTSTRAP_COMMANDS) {
    if (!registry.getCommand(name)) {
      registry.registerCommand(name, undefined, {
        description: '/' + name + ' command'
      })
    }
  }
  stages.push(stageFromElapsed(host, 'commands', 'ok', registry.commandCount + ' commands registered', commandsStarted))

  const toolsStarted = host.monotonicNow()
  if (options.tools?.length) registry.registerFromAgentFunctions(options.tools)
  stages.push(stageFromElapsed(host, 'tools', 'ok', registry.toolCount + ' tools registered', toolsStarted))

  const promptStarted = host.monotonicNow()
  const systemPrompt = buildBootstrapSystemPrompt(context, options.extraContext ?? '', options.tools ?? [])
  stages.push(stageFromElapsed(host, 'system_prompt', 'ok', systemPrompt.length + ' chars', promptStarted))

  return new BootstrapResult({ context, registry, stages, systemPrompt })
}

/** Build the default native prompt without doing filesystem or process I/O. */
export function buildBootstrapSystemPrompt(
  context: Readonly<Record<string, string>>,
  extraContext = '',
  toolDefinitions: readonly unknown[] = [],
): string {
  const tools = bootstrapPromptTools(toolDefinitions)
  const toolNames = new Set(tools.map(tool => tool.name))
  const sections: string[] = [
    'You are Xerxes, an AI coding assistant.',
    '',
    '# Tools available this turn',
  ]
  if (tools.length) {
    for (const tool of tools) {
      const required = requiredFieldSummary(tool.schema)
      sections.push(
        `- ${tool.name}: ${tool.description || 'No description supplied by the host.'}${required ? ` (required: ${required})` : ''}`,
      )
    }
  } else {
    sections.push('- None. Do not emit or simulate tool calls.')
  }
  sections.push(
    '',
    '# How to decide',
    '- Answer directly when knowledge and reasoning are sufficient.',
    '- Do not use tools for greetings, simple arithmetic, or facts you already know.',
    '- Never invoke Python or Node as a calculator.',
    '- Call only tools listed above, and follow each supplied JSON schema exactly.',
    '- If a needed capability is absent, explain the limitation instead of inventing a tool call.',
  )
  if (toolNames.has('Calculator')) {
    sections.push('- Use Calculator only when a calculation genuinely warrants a tool.')
  }
  if (toolNames.has('ReadFile')) {
    sections.push('- Read relevant files with ReadFile before editing them.')
  }
  const writingTools = ['WriteFile', 'FileEditTool', 'AppendFile', 'apply_patch'].filter(name => toolNames.has(name))
  if (writingTools.length) {
    sections.push(`- Available editing tools: ${writingTools.join(', ')}. Choose the narrowest one that fits the change.`)
  }
  if (toolNames.has('exec_command')) {
    const execCommand = tools.find(tool => tool.name === 'exec_command')
    const execProperties = bootstrapRecord(execCommand?.schema?.properties)
    sections.push(execProperties?.args === undefined
      ? '- Use exec_command only when a process is actually needed; follow its provider-supplied schema exactly.'
      : '- exec_command uses direct argv: cmd is one executable and each argument belongs in args; do not put shell syntax in cmd.')
  }
  if (toolNames.has('write_stdin')) {
    sections.push('- Poll or interact with a live terminal session through write_stdin.')
  }
  if (toolNames.has('close_terminal_session')) {
    sections.push('- Close terminal sessions with close_terminal_session when their work is complete.')
  }
  const searchTools = ['GlobTool', 'GrepTool', 'ListDir'].filter(name => toolNames.has(name))
  if (searchTools.length) {
    sections.push(`- Available workspace discovery tools: ${searchTools.join(', ')}.`)
  }
  const webTools = tools.map(tool => tool.name).filter(name => name.startsWith('web.') || name === 'DuckDuckGoSearch')
  if (webTools.length) {
    sections.push(`- Available public-information tools: ${webTools.join(', ')}.`)
  }
  sections.push(
    '',
    '# Context headroom',
    '- Oversized tool results are stored in project agent memory outside model context and replaced with a bounded preview.',
    '- Treat `[Large tool result stored outside model context]` as a valid tool result, not a failure.',
    '- Do not rerun noisy commands just to recover output that is already stored.',
  )
  if (toolNames.has('agent_memory_read')) {
    sections.push('- Read stored-result pointers with agent_memory_read using its supplied schema.')
  }
  if (toolNames.has('SetInteractionModeTool')) {
    sections.push(
      '',
      '# Interaction modes',
      '- code: normal implementation and direct engineering work.',
      '- researcher: read-only exploration, evidence gathering, and cited findings.',
      '- plan: architecture and task breakdown; do not edit code in this mode.',
      '- objective: hard-goal loop for measurable outcomes.',
      '- Switch modes with SetInteractionModeTool and follow its supplied schema.',
      '- Do not final-answer in objective mode while acceptance criteria are unmet; continue iterating or report a concrete blocker with evidence.',
    )
  }
  if (['Agent', 'AgentTool', 'SpawnAgents'].some(name => toolNames.has(name))) {
    sections.push(
      '',
      '# Multi-Agent Orchestration',
      '- Use available agent tools for genuinely independent work and keep task boundaries clear.',
      '- Prefer separate research, implementation, and review paths when parallelism helps.',
    )
  }
  sections.push(
    '',
    '# Critical',
    '- Be concise and direct.',
    '- Respect every tool schema and its workspace/path constraints.',
    '',
    '# Environment',
    '- Date: ' + (context.date ?? ''),
    '- CWD: ' + (context.cwd ?? ''),
    '- Platform: ' + (context.platform ?? ''),
    '- Model: ' + (context.model ?? ''),
  )
  if (context.git_info) sections.push('', '# Git', context.git_info)
  if (context.xerxes_md) sections.push('', '# Project Context', context.xerxes_md)
  if (context.project_agent_workspace) sections.push('', context.project_agent_workspace)
  if (extraContext) sections.push('', extraContext)
  return sections.join('\n')
}

interface BootstrapPromptTool {
  readonly description: string
  readonly name: string
  readonly schema: Readonly<Record<string, unknown>> | undefined
}

function bootstrapPromptTools(definitions: readonly unknown[]): BootstrapPromptTool[] {
  const tools = new Map<string, BootstrapPromptTool>()
  for (const value of definitions) {
    const record = bootstrapRecord(value)
    const functionRecord = bootstrapRecord(record?.function)
    const name = bootstrapString(functionRecord?.name) ?? bootstrapString(record?.name)
    if (!name) continue
    const description = boundedToolDescription(
      bootstrapString(functionRecord?.description) ?? bootstrapString(record?.description) ?? '',
    )
    const schema = bootstrapRecord(functionRecord?.parameters)
      ?? bootstrapRecord(record?.input_schema)
      ?? bootstrapRecord(record?.parameters)
    tools.set(name, { name, description, schema })
  }
  return [...tools.values()]
}

function requiredFieldSummary(schema: Readonly<Record<string, unknown>> | undefined): string {
  if (!Array.isArray(schema?.required)) return ''
  const fields = schema.required
    .filter((value): value is string => typeof value === 'string' && value.trim().length > 0)
    .map(value => value.trim().slice(0, 40))
  const shown = fields.slice(0, 8)
  return shown.join(', ') + (fields.length > shown.length ? `, +${fields.length - shown.length} more` : '')
}

function boundedToolDescription(value: string): string {
  const normalized = value.replace(/\s+/g, ' ').trim()
  return normalized.length <= 240 ? normalized : normalized.slice(0, 237) + '...'
}

function bootstrapRecord(value: unknown): Record<string, unknown> | undefined {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? value as Record<string, unknown>
    : undefined
}

function bootstrapString(value: unknown): string | undefined {
  return typeof value === 'string' && value.trim() ? value.trim() : undefined
}

/** Create the Bun-backed host used by the public convenience entry point. */
export function createSystemBootstrapHost(
  environment: Record<string, string | undefined> = process.env
): BootstrapHost {
  return {
    cwd: () => process.cwd(),
    date: () => new Date(),
    gitInfo: collectGitInfo,
    monotonicNow: () => performance.now(),
    platform: () => process.platform,
    projectWorkspace: async cwd => (await loadProjectAgentWorkspace(cwd)).prompt,
    readText: readOptionalText,
    runtimeVersion: () => 'Bun ' + Bun.version,
    xerxesHomeFile: name => xerxesSubdirFor(environment, name)
  }
}

async function loadXerxesMd(cwd: string, host: BootstrapHost): Promise<string> {
  const parts: string[] = []
  const globalPath = host.xerxesHomeFile('XERXES.md')
  const global = await optionalContext(() => host.readText(globalPath))
  if (global) {
    parts.push('[Global XERXES.md]\n' + scanContextContent(global, 'Global XERXES.md'))
  }

  let current = cwd
  for (let index = 0; index < 10; index += 1) {
    const candidate = join(current, 'XERXES.md')
    const project = await optionalContext(() => host.readText(candidate))
    if (project) {
      const label = 'Project XERXES.md: ' + candidate
      parts.push('[' + label + ']\n' + scanContextContent(project, label))
      break
    }
    const parent = dirname(current)
    if (parent === current) break
    current = parent
  }
  return parts.join('\n\n')
}

/** Collect bounded repository context while running independent Git probes concurrently. */
export async function collectGitInfo(cwd: string, run: GitCommandRunner = runGit): Promise<string> {
  const [branch, rawStatus, log] = await Promise.all([
    run(['rev-parse', '--abbrev-ref', 'HEAD'], cwd),
    run(['status', '--short'], cwd),
    run(['log', '--oneline', '-5'], cwd)
  ])
  if (!branch) return ''
  const status = clipBootstrapGitStatus(rawStatus)
  const parts = ['Branch: ' + branch]
  if (status) parts.push('Status:\n' + status)
  if (log) parts.push('Recent commits:\n' + log)
  return parts.join('\n')
}

/** Clip a potentially huge porcelain listing without splitting a UTF-8 path or hiding truncation. */
export function clipBootstrapGitStatus(status: string): string {
  if (Buffer.byteLength(status, 'utf8') <= MAX_BOOTSTRAP_GIT_STATUS_BYTES) {
    return status
  }

  const lines = status.split('\n').filter(line => line.length > 0)
  const kept: string[] = []
  let usedBytes = 0

  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index]
    if (line === undefined) continue
    const separatorBytes = kept.length > 0 ? 1 : 0
    const candidateBytes = usedBytes + separatorBytes + Buffer.byteLength(line, 'utf8')
    const remaining = lines.length - index - 1
    const markerBytes = remaining ? 1 + Buffer.byteLength(gitStatusOmissionMarker(remaining), 'utf8') : 0
    if (candidateBytes + markerBytes > MAX_BOOTSTRAP_GIT_STATUS_BYTES) {
      break
    }
    kept.push(line)
    usedBytes = candidateBytes
  }

  const omitted = lines.length - kept.length
  const marker = gitStatusOmissionMarker(omitted)
  return kept.length > 0 ? kept.join('\n') + '\n' + marker : marker
}

function gitStatusOmissionMarker(omitted: number): string {
  return `[truncated: ${omitted} additional changes omitted; run git status --short for the complete list]`
}

async function runGit(arguments_: readonly string[], cwd: string): Promise<string> {
  try {
    const child = Bun.spawn(['git', ...arguments_], {
      cwd,
      stdin: 'ignore',
      stdout: 'pipe',
      stderr: 'pipe'
    })
    const [exitCode, stdout] = await Promise.all([child.exited, new Response(child.stdout).text()])
    return exitCode === 0 ? stdout.trim() : ''
  } catch {
    return ''
  }
}

async function readOptionalText(path: string): Promise<string | undefined> {
  try {
    return await readFile(path, 'utf8')
  } catch {
    return undefined
  }
}

async function optionalContext(read: () => Promise<string | undefined>): Promise<string> {
  try {
    return (await read()) ?? ''
  } catch {
    return ''
  }
}

function formatBootstrapDate(date: Date): string {
  const year = date.getFullYear().toString().padStart(4, '0')
  const month = (date.getMonth() + 1).toString().padStart(2, '0')
  const day = date.getDate().toString().padStart(2, '0')
  const weekday = new Intl.DateTimeFormat('en-US', { weekday: 'long' }).format(date)
  return year + '-' + month + '-' + day + ' ' + weekday
}

function stageFromElapsed(
  host: BootstrapHost,
  name: string,
  status: BootstrapStageStatus,
  detail: string,
  started: number
): BootstrapStage {
  return {
    name,
    status,
    detail,
    durationMs: Math.max(0, host.monotonicNow() - started)
  }
}
